HW2/P2/P2.txt

```diff
@@ -0,0 +1,24 @@
+For fine grained lock and medium grained lock, it takes much more time than serial version in both correlated and uncorrelated case, due to a lot of overhead. Here is one result with N=10 for medium grained lock:
+
+Serial uncorrelated: 0.245481014252 seconds
+Fine grained uncorrelated: 4.79588294029 seconds
+Medium grained uncorrelated: 5.32975411415 seconds
+Serial correlated: 0.272443056107 seconds
+Fine grained correlated: 4.33650183678 seconds
+Medium grained correlated: 4.09460902214 seconds 
+
+The time for medium grained locking on correlated data is expectedly less than that on uncorrelated data, because the src and dest are no more than 10 elements away, so when I lock with 10 adjacent elements, I only need to lock once for both src and dest, thus reduce the time.
+
+To find an appropriate N, I tested N in range(1,50,3), and get the image P2.png.
+In the image, the N didn't seem to reduce the time for uncorrelated data, but for correlated data, it seems N=37 would be a good choice. But note that the time is still much more than serial version.
```

Comment:
I found that N=20 was the best due to the elements not being more than 10 apart. I wonder if doing more trials and taking the averages would give a more consistent result since N=37 seems rather arbitrary